Sources/SwiftLexicalLookup/Scopes/ScopeImplementations.swift

```diff
@@ -348,27 +459,107 @@ import SwiftSyntax
+  @_spi(Experimental) public var scopeDebugName: String {
+    "ActorDeclScope"
+  }
+}
+@_spi(Experimental) extension ClassDeclSyntax: NominalTypeDeclSyntax {
+  @_spi(Experimental) public var scopeDebugName: String {
+    "ClassDeclScope"
+  }
+}
+@_spi(Experimental) extension StructDeclSyntax: NominalTypeDeclSyntax {
+  @_spi(Experimental) public var scopeDebugName: String {
+    "StructDeclScope"
+  }
+}
+@_spi(Experimental) extension EnumDeclSyntax: NominalTypeDeclSyntax {
+  @_spi(Experimental) public var scopeDebugName: String {
+    "EnumDeclScope"
+  }
+}
+@_spi(Experimental) extension ExtensionDeclSyntax: LookInMembersScopeSyntax {
+  @_spi(Experimental) public var lookupMembersPosition: AbsolutePosition {
+    extendedType.position
+  }
+
+  @_spi(Experimental) public var introducedNames: [LookupName] {
+    []
+  }
+
+  @_spi(Experimental) public var scopeDebugName: String {
+    "ExtensionDeclScope"
+  }
+
+  @_spi(Experimental) public func lookup(
+    _ identifier: Identifier?,
+    at lookUpPosition: AbsolutePosition,
+    with config: LookupConfig
+  ) -> [LookupResult] {
+    if memberBlock.range.contains(lookUpPosition) {
+      return [.lookInMembers(self)] + defaultLookupImplementation(identifier, at: lookUpPosition, with: config)
+    } else {
+      return defaultLookupImplementation(identifier, at: lookUpPosition, with: config)
+    }
+  }
+}
 
 @_spi(Experimental) extension AccessorDeclSyntax: ScopeSyntax {
   /// Implicit and/or explicit names introduced within the accessor.
   @_spi(Experimental) public var introducedNames: [LookupName] {
+    let names: [LookupName]
+
     if let parameters {
-      return LookupName.getNames(from: parameters)
+      names = LookupName.getNames(from: parameters)
     } else {
       switch accessorSpecifier.tokenKind {
       case .keyword(.set), .keyword(.willSet):
-        return [.implicit(.newValue(self))]
+        names = [.implicit(.newValue(self))]
       case .keyword(.didSet):
-        return [.implicit(.oldValue(self))]
+        names = [.implicit(.oldValue(self))]
       default:
-        return []
+        names = []
       }
     }
+
+    return names
+  }
+
+  @_spi(Experimental) public var scopeDebugName: String {
+    "AccessorDeclScope"
+  }
+
+  /// Returns result with matching names from
+  /// this scope and passes result with implicit `self`
+  /// to be introduced after the `subscript`
+  /// declaration scope grandparent.
+  @_spi(Experimental) public func lookup(
+    _ identifier: Identifier?,
+    at lookUpPosition: AbsolutePosition,
+    with config: LookupConfig
+  ) -> [LookupResult] {
+    guard let parentAccessorBlockScope = parentScope?.as(AccessorBlockSyntax.self) else {
+      return defaultLookupImplementation(identifier, at: lookUpPosition, with: config)
+    }
+
+    let implicitSelf: [LookupName] = [.implicit(.self(self))]
+      .filter { name in
+        checkIdentifier(identifier, refersTo: name, at: lookUpPosition)
+      }
```

Comment:
Should we only add `self` if this is a member? Ie. we shouldn’t add `self` in a top level variable with an accessor, right?

Reply:
This case was handled here implicitly. The result with `self` was handed over to a special function that passed it to `AccessorBlockSyntax` parent which later handed it off to it’s `SubscriptDeclSyntax` parent. 

I’ve noticed now this solution wasn’t perfect though. It was quite inflexible and didn’t work for variable declaration accessors. To fix this, I introduced now a new specialized scope kind: [`CanInterleaveResultsLaterScopeSyntax`](https://github.com/swiftlang/swift-syntax/blob/433f14bbd85810efd88954e0cb62c3b655ca7037/Sources/SwiftLexicalLookup/Scopes/CanInterleaveResultsLaterScopeSyntax.swift#L15-L22) with `lookupWithInterleavedResults` function to further generalize this behavior. [Now accessor can hand the result with `self` keyword to it’s `CanInterleaveResultsLaterScopeSyntax` parent](https://github.com/swiftlang/swift-syntax/blob/433f14bbd85810efd88954e0cb62c3b655ca7037/Sources/SwiftLexicalLookup/Scopes/ScopeImplementations.swift#L556-L561). The `self` keyword is introduced properly by accessors inside `subscript` and variables and it does differentiate whether it’s a member or not.